TACHO_DRIVER.md

```diff
@@ -0,0 +1,199 @@
+determines how the motor slows down. If the new setpoint is towards a
+higher absolute value, then the `ramp_up_sp` determines how the motor
+speeds up.
+
+An example follows:
+
+- `ramp_up_sp` is set to 4000
+- `ramp_down_sp` is set to 2000
+- `speed_regulation` is set to `off`
+- `duty_cycle_sp` is set to 75%
+
+If we write `run-direct` to the motor `command` attribute, the motor will
+take 3000 msec to ramp up to 75% duty cycle. What happens when we change the
+`duty_cycle_sp` to other values while the motor is running?
+
+| New `duty_cycle_sp` | Operation
+| ------------------- | ----------------------
+|   100               | Motor ramps up to 100% in 1000 msec
+|   50                | Motor ramps down to 50% in 1000 msec
+|   -50               | Motor ramps down to 0% in 1000 msec and up to -50 in 2000 msec
+
+## Run Modes
+
+The `run` attribute determines function that the motor is executing. Within a `run`
+mode, some attributes have an effect, while others do not. These details are
+described below.
+
+The run modes can be classified into two types. The `run-direct` is suited to
+applications where the programmer wants direct control of the motors. The other
+three modes:
+
+- `run_timed`
+- `run_to_abs_pos`
+- `run_to_rel_pos`
+
+They can be through of as convenience commands where the driver handles the motor
+control independently of the application. This is well suited to applications where you
+want the driver to handle the details, like machines where the motors must alternate
+between precise positions or run for fixed timing periods.
+
+### `run-direct` Mode
+
+In its most basic operating mode, the `tacho-motor` accepts the `run-direct`
+command. After accepting a `run-direct` command, the motor will only accept
+a `stop` command - the other commands represent "helpers" that automatically
+do what `run-direct` is capable of without user intervention.
+
+NOTE: In this mode the `stop` command stops the motor immediately! If you
+want the motor to use the ramping feature, change the desired setpoint to 0.
+
+Currently, only changes to the `duty_cycle_sp` take effect
+immediately and I'm proposing that this should depend on the `speed_regulation`
+setting. I may be convinced to also allow changes to the ramp setpoints.
+
+You cannot change the `speed_regulation` value without stopping the
+motor first. (Not sure on this - I can be convinced to change this)
+
+Finally, if the ramping target setpoint has been set to 0, the motor is
+stopped in the corresponding `stop_mode`.
+
+### `run-forever` Mode
+
+Deprecated? This does not provide any new features that are not already present
+in `run-direct`
+
+### `run-timed`, `run_to_abs_pos` and `run_to_rel_pos` Modes
+
+These are the convenience commands described earlier. The motors will not
+respond to changes in any setpoints or command except for `stop`. The `stop`
+command turns the motor off immediately in whatever the `stop_mode` is set
+to - no ramping will take place.
```

Comment:
Ramp down would not be possible with other run commands if `stop` did not trigger the ramp. This is also why I'm thinking we don't try to implement ramp down in the `run-direct` command.

Another thing to keep in mind that an important difference between `run-direct` command `dc-motor` class is that you can still read the `position` and `state` attributes. This is another reason why I'm am in favor of `run-direct` being strictly for "implement your own PID". If you just want to treat the motor like a `dc-motor`, then you can switch the output port to `dc-motor` mode and use the `dc-motor` class.